File: app/src/main/java/com/scop/org/minesweeper/generators/RandomGenerator.java
package com.scop.org.minesweeper.generators;

import com.scop.org.minesweeper.elements.Grid;
import com.scop.org.minesweeper.elements.Tile;
import com.scop.org.minesweeper.utils.GridUtils;

import java.util.List;
import java.util.Random;

public class RandomGenerator implements GridGenerator {

	@Override
	public void generateNewGrid(Grid grid, int bombs, FinishCallback cb) {
		new Thread(){
			@Override
			public void run(){
				generateNewRandomGrid(grid, bombs);
				cb.finished();
			}
		}.start();
	}

	protected void generateNewRandomGrid(Grid grid, int bombs) {
		int w = grid.getW();
		int h = grid.getH();
		List<Tile> tiles = grid.getGrid();
		tiles.clear();


		for (int j=0; j<h; j++) {
			for (int i=0; i<w; i++) {
				tiles.add( new Tile(i,j,Tile.Status.COVERED) );
			}
		}

		Random random = new Random();
		int idx,max = w*h;
		for (int i=0;i<bombs;i++) {
			idx = random.nextInt(max);
			if (tiles.get(idx).hasBomb()){
				i--;
				continue;
			}
			tiles.get(idx).plantBomb();

			GridUtils.getNeighbors(grid, tiles.get(idx)).forEach(Tile::hasBombNear);
		}
	}
}

File: app/src/main/java/org/oar/minesweeper/generators/RandomGenerator.kt
package org.oar.minesweeper.generators

import org.oar.minesweeper.models.Grid
import org.oar.minesweeper.models.GridGenerationDetails
import org.oar.minesweeper.models.Tile
import org.oar.minesweeper.models.TileStatus
import org.oar.minesweeper.utils.GridUtils.getNeighbors
import java.util.*

open class RandomGenerator : GridGenerator {
    override fun generateNewGrid(grid: Grid, onFinish: (GridGenerationDetails) -> Unit) {
        Thread {
            generateNewRandomGrid(grid)
            onFinish(GridGenerationDetails())
        }.start()
    }

    protected fun generateNewRandomGrid(grid: Grid) {
        val w = grid.width
        val h = grid.height
        val bombs = grid.bombs

        grid.tiles.clear()
        for (j in 0 until h) {
            for (i in 0 until w) {
                grid.tiles.add(Tile(i, j, TileStatus.COVERED))
            }
        }

        generateBombs(grid, bombs)
    }

    override fun forceCleanSpot(grid: Grid, tile: Tile) {
        val neighbors = grid.getNeighbors(tile)

        val defusedBombs = tile.bombsNear + if (tile.hasBomb) 1 else 0

        tile.hasBomb = false
        tile.bombsNear = 0

        val skipTiles = neighbors
            .onEach { it.hasBomb = false }
            .onEach { neighbor ->
                val level2Neighbors = grid.getNeighbors(neighbor)
                neighbor.bombsNear = level2Neighbors.count { it.hasBomb }

                level2Neighbors.forEach {
                    it.bombsNear = grid.getNeighbors(it).count { i -> i.hasBomb }
                }
            }
            .map { it.x + it.y * grid.width }
            .toMutableList()
            .apply { add(tile.x + tile.y * grid.width) }

        generateBombs(grid, defusedBombs, skipTiles)
    }

    private fun generateBombs(grid: Grid, bombs: Int, skipTiles: List<Int> = listOf()) {
        if (bombs == 0) return

        val w = grid.width
        val h = grid.height

        val random = Random()
        var idx: Int
        val max = w * h
        var i = 0
        while (i < bombs) {
            idx = random.nextInt(max)
            if (grid.tiles[idx].hasBomb || skipTiles.contains(idx))
                continue

            grid.tiles[idx].hasBomb = true
            grid.getNeighbors(grid.tiles[idx]).forEach { it.bombsNear++ }
            i++
        }
    }
}